### ISSUE TITLE
[JENKINS-30231] Build creates second workspace@2 for non-concurrent build configuration

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#880]
<p>I didn't have this issue when I was using ver 1.580.3.<br/>
After upgrading Jenkins ver 1.609.1 (together with other Jenkins plugins update),<br/>
I started to see this issue.</p>

<p>My jenkins job is using multi-config project type (aka Matrix project),<br/>
and has 3 variants.<br/>
I didn't set to be run concurrently,<br/>
so the 3 variants will start the build one by one.</p>

<p>In the previous version 1.580.3,<br/>
it will reuse the same workspace folder for all of 3 variant builds.</p>

<p>In the new version 1.609.1,<br/>
when multiple jobs are triggerd,<br/>
I can see sometimes the first variant build of the second job<br/>
will start to create and use workspace@2<br/>
instead of re-using original worksapce.</p>

<p>Note:<br/>
I'm also using disk-usage plugin.<br/>
At the end of build,<br/>
it will trigger that plugin to calculate workspace disk space.<br/>
My work space tends to be large and sometimes it take several minutes to finish.</p>

<p>I suspect it's that disk-usage plugin still occupying the original worksapce,<br/>
but the second job in queue starts earlier,<br/>
so it creates workspace@2 for the first variable build.</p>

<p>From workspace@2,<br/>
I always see only one variant only under AXIS folder.</p>

**Attachments:**

- [image-2020-08-16-13-41-37-491.png](https://issues.jenkins.io/secure/attachment/52222/image-2020-08-16-13-41-37-491.png)
- [image-2020-08-16-13-42-26-759.png](https://issues.jenkins.io/secure/attachment/52223/image-2020-08-16-13-42-26-759.png)
- [image-2020-08-16-13-43-23-877.png](https://issues.jenkins.io/secure/attachment/52224/image-2020-08-16-13-43-23-877.png)

---
<details><summary><i>Originally reported by <a title="Rick Liu" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=totoroliu">totoroliu</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-30231" target="_blank">Build creates second workspace@2 for non-concurrent build configuration</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Devin Nusbaum" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum">dnusbaum</a>
<li><b>reported by</b>: totoroliu
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2018-09-24T13:03:46+00:00
<li><b>votes</b>: 4
<li><b>watchers</b>: 22
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=totoroliu">totoroliu</a>:</i>
<p>Per comment from Jesse Glick,<br/>
I create this separate issue<br/>
which has similar symptom.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=will177">will177</a>:</i>
<p>I am seeing this issue today on one of my pipelines.  Is there a quick fix workaround I can use to get rid of it and reset the pipeline?  I don't mind starting the pipeline from scratch again - just wondering how to get rid of the @2 workspace.  Anyone know?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mksmith">mksmith</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=will177" class="user-hover" rel="will177">will177</a> It looks like this might be a solution: <a href="https://issues.jenkins-ci.org/browse/JENKINS-14354" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-14354</a></p>

<p>It looks like you can replace the "@" separator with a separator of your choice. </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=totoroliu">totoroliu</a>:</i>
<p>I use customize worksapce configuration to hardcoded the workspace name to work-around this problem</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rg">rg</a>:</i>
<p>This is still a problem. We restrict our jobs to one per node and still see a number of @2 workspaces. Replacing @ with _ (like <a href="https://issues.jenkins.io/browse/JENKINS-14354" title="Concurrent build jobs use BAD naming convention for workspaces" class="issue-link" data-issue-key="JENKINS-14354"><del>JENKINS-14354</del></a>) helps builds which get tripped up by @ but we still have duplication. As we have a number of large workspaces this wastes a lot of diskspace.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ndza2017">ndza2017</a>:</i>
<p>We've recently started seeing a duplicate @2 workspace as well, when we moved from 2.60.3 to 2.73.2 with some new plug in.</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum">dnusbaum</a>:</i>
<p><a href="https://issues.jenkins.io/browse/JENKINS-41127" title="Multiple pipeline instances running concurrently when concurrent execution disabled" class="issue-link" data-issue-key="JENKINS-41127"><del>JENKINS-41127</del></a>, which may have been the root cause of this issue, was just fixed in <a href="https://jenkins.io/changelog/#v2.136" class="external-link" target="_blank" rel="nofollow noopener">Jenkins 2.136</a>. Does anyone have a test environment in which they can consistently reproduce this issue? If so, it would be great if you could test whether this problem still occurs in Jenkins 2.136 or newer.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum">dnusbaum</a>:</i>
<p>Since I have not heard back either way, I am going to go ahead and close this issue because at least one way it could happen was fixed in Jenkins 2.136. Please feel free to reopen the issue if you are still seeing it in Jenkins 2.136 or newer.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kamil02">kamil02</a>:</i>
<p>Hi! </p>

<p>Yesterday one of my jobs started to show the same behavior, but we are not using the matrix-project-plugin. At some point (maybe after we deleted the original workspace-folder of a certain job) jenkins started to build the project in a folder suffixed by "@2" - but only on one specific node. Recreating the Pipelinejob with the same name didn't help. Today i did a restart of the server and that seems to have soled the issue for now.</p>

<p>We are using Jenkins 2.138.2.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kamil02">kamil02</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=j1mehta" class="user-hover" rel="j1mehta">j1mehta</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> I think we triggered this issue by removing the working directory of the job within the workspace of a node by hand. </p>

<p>Something like </p>
<ol>
	<li>Run job_a on node b</li>
	<li>let it finish</li>
	<li>login via ssh onto node b</li>
	<li>rm -R {JENKINS_WORKSPACE}/job_a</li>
	<li>rerun job a</li>
	<li>new folder with @2 suffix should be created</li>
</ol>


<p>Unfortunately i cant evaluate this right now.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kamil02">kamil02</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=j1mehta" class="user-hover" rel="j1mehta">j1mehta</a> did you also try to restart the jenkins master after the deletion of both (original and @2) workspaces? If i remember correctly, i also had the feeling, that jenkins uses some kind of cache to remember where the workspace is supposed to be. As mentioned before, restarting the jenkins master instance after deleting the original and @2 workspace solved our problem.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=j1mehta">j1mehta</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kamil02" class="user-hover" rel="kamil02">kamil02</a>, I don't have admin access to restart Jenkins master. Besides, thats only a temporary workaround. If this was a permanent solution, it shouldn't have cropped up for a new project (or same project with its name changed for that matter). It is still a bug. Will try to come up with a more consistent way to produce it. </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=adamkolarik">adamkolarik</a>:</i>
<p>I'm seeing this issue on version 2.176.1. We also don't have concurrent or Matrix builds setup, just one executor per node. On one specific node, we got the workspace@2 duplicated workspace. I tried deleting both workspaces on the node manually and restart Jenkins server to see if that helps and it did.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kamil02">kamil02</a>:</i>
<p>Today the same error occurred again: Jenkins created a ..@2 and ...@3 folder in the .../workspace directory. I had to delete these via SSH and restart the master => back to normal job workspace again.  </p>

<p>Jenkins ver. 2.176.3</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=xianpeng">xianpeng</a>:</i>
<p>I have the same issue on Jenkins ver. 2.176.3.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=naveen6s">naveen6s</a>:</i>
<p>I have the same issue with Jenkins version 2.222.3</p>

<p>I have admin access, so I rebooted Jenkins master after deleting ..@2 folder. This did not help either.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olssons">olssons</a>:</i>
<p>I also have this problem.  I checkout two repositories into my workspace and often times Jenkins creates a second workspace for the parallel-running job which checks out the second repo.  The tools assume they are in the same workspace and promptly fail.  Possible workaround might be to save the location of WORKSPACE in the beginning and then with each parallel task, always change directory (cd) back to that location before doing anything.  </p>

<p>For example,</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/52223/52223_image-2020-08-16-13-42-26-759.png" height="154" width="509" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/52224/52224_image-2020-08-16-13-43-23-877.png" height="209" width="728" style="border: 0px solid black" /></span></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olssons">olssons</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> - I see that the issue is marked as Resolved.  In this case, should a new issue be created or the preference is to Reopen?  I didn't want to reopen if it really needs to be a new issue instead.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum">dnusbaum</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olssons" class="user-hover" rel="olssons">olssons</a> My guess is that there are probably many distinct issues being reported here (e.g. some scenarios are likely specific to (multibranch) pipelines, but there were also bugs fixed that affected freestyle and other project types). I would go ahead and open a new issue with full details on your specific case, ideally with steps on how to reproduce the problem from scratch.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olssons">olssons</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dnusbaum" class="user-hover" rel="dnusbaum">dnusbaum</a> - With my issue, the double workspace never occurs during the "Build Software" parallel tasks.  However, when I have 3 parallel tasks in a row, like under "Analysis and Adaptation", this is where the problem <b>sometimes</b> happens.  I say <b>sometimes</b> because it happens maybe half the time or 3 out of 4 builds in a row. </p>

<p>I first started my workarounds by removing the second checkout like I mentioned before (see "Checkout tools" above).  That did not fix the problem.  I then made all the tasks in Analysis and Adaptation serial.  That (obviously) fixed the problem.  If I were to guess, it didn't like 3 tasks running in parallel, all using the same workspace.  My builds take about 10-15 minutes longer now, but no more failures at least.</p>

<p>I may open an issue in the future. </p>

<p>PS - I could never recreate the problem from the Jenkins Pipeline script, only from a checked in Jenkinsfile itself</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wizzzozzz">wizzzozzz</a>:</i>
<p>Is there anyway to disable this option?</p>

<p>so no new workspace will be created and only the original workspace will be used alone? (I'm aware of the risks)</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-21622">JENKINS-21622</a>]</i>
